scala-package/core/src/main/scala/org/apache/mxnet/FeedForward.scala

```diff
@@ -224,13 +224,24 @@ class FeedForward private(
     var i = 0
     while (data.hasNext && i != numBatch) {
       val batch = data.next()
-      i += 1
-      ExecutorManager.loadData(batch, dataArrays)
-      predExec.forward(isTrain = false)
-      val padded = batch.pad
-      val realSize = batchSize - padded
-      for ((list, nd) <- outputs zip predExec.outputs) {
-        list += nd.slice(0, realSize).copy()
+      try {
+        i += 1
+        ExecutorManager.loadData(batch, dataArrays)
+        predExec.forward(isTrain = false)
+        val padded = batch.pad
+        val realSize = batchSize - padded
+        for ((list, nd) <- outputs zip predExec.outputs) {
+          // The slice is being written to a value so that dispose can be called after the copy.
+          // The one liner nd.slice().copy() leads to leaking the memory of the slice.
+          val ndSliced = nd.slice(0, realSize)
+          try {
+            list += ndSliced.copy()
+          } finally {
+            ndSliced.dispose()
+          }
+        }
+      } finally {
+        batch.dispose()
```

Comment:
shouldn't it be the dataArrays that needs to be disposed ? see [loadData](https://github.com/apache/incubator-mxnet/blob/590ec3f508645ae06960a93293193f6280ed1c63/scala-package/core/src/main/scala/org/apache/mxnet/ExecutorManager.scala#L254) and [loadDatageneral](https://github.com/apache/incubator-mxnet/blob/590ec3f508645ae06960a93293193f6280ed1c63/scala-package/core/src/main/scala/org/apache/mxnet/ExecutorManager.scala#L227) where its doing a copy from batch to dataArrays

I think we shouldn't dispose the original input from the Iterator, for example if the user wants to use the same input on an another model

Reply:
We aren't disposing of the original input. When data.next() is called, a slice of the data is made [here](https://github.com/apache/incubator-mxnet/blob/master/scala-package/core/src/main/scala/org/apache/mxnet/io/NDArrayIter.scala#L163). This slice is what should be getting disposed of when we dispose of batch.

Reply:
Thanks Andrew, what about dataArrays?

Reply:
Summarizing the conversation we had offline: It appears that the copyTo in loadData is copying to memory in the Executor predExec. The memory in the executor seems to get reused across predict calls. I think that the proper place to dispose of this part would be to add a dispose method for FeedForward and have that dispose of it's predExec.